### PR TITLE
[protobuf] update to 3.22.2

### LINKED ISSUE #29697: [protobuf] update to 3.22.2
Library name: protobuf

New version number: 3.22.2

Other information that may be useful (release notes, etc...)
https://github.com/protocolbuffers/protobuf/releases/tag/v3.22.0

### PR DESCRIPTION
**Pull request by jimwang118**
Fixes https://github.com/microsoft/vcpkg/issues/29697
Because compiling `protoc` needs to rely on the header files in `abseil`, the dependency on `abseil` is added.
The content of the `compile_options.patch` patch file has been recently fixed in the upstream code, so delete the patch file. 
https://github.com/protocolbuffers/protobuf/blob/v22.2/CMakeLists.txt#:~:text=with%20multiple%20processes-,add_compile_options(/MP),add_compile_options(,-/wd4065%20%23%20switch

- [X] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md)
- [X] SHA512s are updated for each updated download
- [ ] ~~The "supports" clause reflects platforms that may be fixed by this new version~~
- [ ] ~~Any fixed [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) entries are removed from that file.~~
- [X] Any patches that are no longer applied are deleted from the port's directory.
- [X] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [X] Only one version is added to each modified port's versions file.

### CODE REVIEW

File: ports/protobuf/portfile.cmake [hunk 1]
```diff
@@ -46,7 +46,7 @@ vcpkg_cmake_configure(
 )
 
 vcpkg_cmake_install()
-
+vcpkg_cmake_config_fixup(CONFIG_PATH lib/cmake/utf8_range)
```
Comment (dg0yt) [inline ports/protobuf/portfile.cmake:49]: This doesn't look right for port protobuf.

Comment (jimwang118) [reply ports/protobuf/portfile.cmake:49]: Upstream has added `uft8_range `compilation and installation scripts (https://github.com/protocolbuffers/protobuf/blob/main/CMakeLists.txt#:~:text=if%20(protobuf_BUILD_PROTOBUF_BINARIES),cmake/utf8_range.cmake)), and linked the library in `utf8_range `when compiling `libprotobuf-lite` and `libprotobuf`.(https://github.com/protocolbuffers/protobuf/blob/main/cmake/libprotobuf-lite.cmake#:~:text=target_include_directories(libprotobuf%2Dlite%20PRIVATE,libprotobuf%2Dlite%20PRIVATE%20utf8_validity))

Comment (dg0yt) [reply ports/protobuf/portfile.cmake:49]: Maybe, but ...
Does this need to be devendored?
Is utf8_range cmake config found when installed to `share/protobuf`?

**Comment by coryan** [comment]
I think this should be 4.22.2 for C++.  Protobuf versioning is a bit ad-hoc.  Each language (in this case C++) has its own major version:

https://protobuf.dev/news/2022-05-06/

The v22.x series is a breaking change for C++:

https://protobuf.dev/news/2022-08-03/
https://github.com/protocolbuffers/protobuf/releases/tag/v22.0

Therefore the v22.x series is 4.22.x for C++.  The protobuf creates tags (at the same SHA) with the right version for each language:

https://github.com/protocolbuffers/protobuf/releases/tag/v22.2
https://github.com/protocolbuffers/protobuf/releases/tag/v4.22.2
https://github.com/protocolbuffers/protobuf/releases/tag/v3.22.2

File: ports/protobuf/portfile.cmake (hunk 1 above)
Comment (coryan) [reply ports/protobuf/portfile.cmake:49]: FWIW, the Protobuf and gRPC teams are working on this:

https://github.com/protocolbuffers/protobuf/issues/12185

**Comment by talregev** [comment]
I didn't understand the versioning of protobuf.
https://protobuf.dev/support/version-support/
in C++ there is 2 versions:
23.x | 4.23.x
Not sure what the different.

**Comment by coryan** [comment]
> I didn't understand the versioning of protobuf. https://protobuf.dev/support/version-support/ in C++ there is 2 versions: 23.x | 4.23.x Not sure what the different.

The source code for 23.x and 4.23.x is exactly the same.  The first number (23.x) is the version of Protobuf, regardless of the programming language you are using.  The second number (4.23.x) is the version of "Protobuf for C++".

The numbers are different because a new version of Protobuf may be a breaking change for C++, but may not be a breaking change for Python (or any of the other languages Protobuf supports).  In such an scenario, the major version for C++ would increase.

For example, suppose Protobuf 28.x is at "Protobuf for C++ 5.28.x" and "Protobuf for Python 7.28.x".  We assumed Protobuf 29.x is breaking for C++, but not breaking for Python.  Then the version numbers would be "Protobuf 29.x", "Protobuf for C++ 6.29.x" and "Protobuf for Python 7.29.x".

**Comment by talregev** [comment]
@coryan Thank you for detail explanation. so 4.23 is breaking the c++. 
 So we should stay on 3.21.x

**Comment by coryan** [comment]
> @coryan Thank you for detail explanation. so 4.23 is breaking the c++. So we should stay on 3.21.x

You may need to stay on 3.21.x.  However, maybe you should review the list of breaking changes first?  It is possible they do not affect your application.  These links may help:

https://protobuf.dev/news/2022-08-03/#cpp-changes
https://protobuf.dev/support/migration/#cpp-22

**Comment by talregev** [comment]
> > @coryan Thank you for detail explanation. so 4.23 is breaking the c++. So we should stay on 3.21.x
> 
> You may need to stay on 3.21.x.  However, maybe you should review the list of breaking changes first?  It is possible they do not affect your application.  These links may help:
> 
> https://protobuf.dev/news/2022-08-03/#cpp-changes
> https://protobuf.dev/support/migration/#cpp-22
> 

I meant for vcpkg.

**Comment by coryan** [comment]
> > > @coryan Thank you for detail explanation. so 4.23 is breaking the c++. So we should stay on 3.21.x
> > 
> > 
> > You may need to stay on 3.21.x.  However, maybe you should review the list of breaking changes first?  It is possible they do not affect your application.  These links may help:
> > https://protobuf.dev/news/2022-08-03/#cpp-changes
> > https://protobuf.dev/support/migration/#cpp-22
> 
> I meant for vcpkg.

Ah, my mistake.  Something for the vcpkg team to ponder I guess.  I hope they don't plan to freeze the vcpkg ecosystem forever though.

**Comment by coryan** [comment]
> AFAIU the update can be merged when all downstream ports build with it.

One thing to consider: no version of gRPC compiles with Protobuf 22.x (the version proposed in this PR).  gRPC < 1.55 requires Protobuf <= 21.x.  gRPC >= 1.55 requires Protobuf >= 23.x.  Note how it skipped Protobuf 22.x.

And I suspect almost all downstream packages that use Protobuf will need some fixing, at the very least adding `https://cmake.org/cmake/help/latest/variable/CMAKE_FIND_PACKAGE_PREFER_CONFIG.html` because `find_package(Protobuf)` does not work (see https://github.com/protocolbuffers/protobuf/issues/12292 and https://gitlab.kitware.com/cmake/cmake/-/issues/24321).

In addition, `upb` requires an update to get gRPC working, and I could not figure it out in the time I had for this.

**Comment by jimwang118** [comment]
Duplicate of https://github.com/microsoft/vcpkg/pull/32602

**Pull request closed without merging**